controllers/openstackcluster_controller.go

```diff
@@ -627,3 +617,51 @@ func handleUpdateOSCError(openstackCluster *infrav1.OpenStackCluster, message er
 	openstackCluster.Status.FailureReason = &err
 	openstackCluster.Status.FailureMessage = pointer.String(message.Error())
 }
+
+// filterSubnets retrieves the subnets based on the Subnet filters specified on OpenstackCluster.
+func filterSubnets(networkingService *networking.Service, openStackCluster *infrav1.OpenStackCluster) ([]infrav1.Subnet, error) {
+	var subnets []infrav1.Subnet
+	openStackClusterSubnets := openStackCluster.Spec.Subnets
+	if len(openStackClusterSubnets) == 0 {
+		empty := &infrav1.SubnetFilter{}
+		listOpt := empty.ToListOpt()
+		listOpt.NetworkID = openStackCluster.Status.Network.ID
+		filteredSubnets, err := networkingService.GetSubnetsByFilter(listOpt)
+		if err != nil {
+			err = fmt.Errorf("failed to find subnets: %w", err)
+			if errors.Is(err, networking.ErrFilterMatch) {
+				handleUpdateOSCError(openStackCluster, err)
+			}
+			return nil, err
+		}
+		if len(filteredSubnets) > 2 {
+			return nil, fmt.Errorf("more than two subnets found in the Network. Specify the subnets in the OpenStackCluster.Spec instead")
+		}
+		for i := range filteredSubnets {
+			subnets = append(subnets, infrav1.Subnet{
+				ID:   filteredSubnets[i].ID,
+				Name: filteredSubnets[i].Name,
+				CIDR: filteredSubnets[i].CIDR,
+				Tags: filteredSubnets[i].Tags,
+			})
+		}
+	} else {
+		for subnet := range openStackClusterSubnets {
```

Comment:
Same here.

Reply:
I'm iterating like this to avoid hitting the issue that was explained here https://stackoverflow.com/questions/62446118/implicit-memory-aliasing-in-for-loop/68247837#68247837

Reply:
Makes sense, thanks!